### PR TITLE
feat: serialization support for StateBag and NamespacedStateBag

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request adds comprehensive JSON and YAML serialization support to the `StateBag` and `NamespacedStateBag` types in the `automa` package, making it possible to persist, transmit, or restore workflow state in a predictable way. It also includes extensive documentation and tests to guide users on best practices and potential pitfalls when serializing state.

**Serialization support for StateBag and NamespacedStateBag:**

- Implemented `json.Marshaler`, `json.Unmarshaler`, `yaml.Marshaler`, and `yaml.Unmarshaler` interfaces for `SyncStateBag` and `SyncNamespacedStateBag`, enabling direct serialization and deserialization to/from JSON and YAML. The serialization maps `Key` values to strings and snapshots the bag contents under appropriate locks. [[1]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR104-R167) [[2]](diffhunk://#diff-6800f18f380946d94dd739a46e1ce72d445cf99d55bc2e42fb5c58b43284326bR183-R310)
- Added a helper function `StateBagToStringMap` to convert a `StateBag`'s items to a `map[string]interface{}` for use in serialization.
- Updated the `StateBag` interface to require these serialization methods, ensuring all implementations support JSON/YAML encoding.

**Documentation and guidance:**

- Added a new documentation page, `state-serialization.md`, detailing the serialization format, type mapping, recommended usage, and common pitfalls (such as numeric type conversion in JSON).
- Linked the new documentation from the main README.

**Testing:**

- Added comprehensive round-trip tests for both `SyncStateBag` and `SyncNamespacedStateBag` covering JSON and YAML serialization, type conversions, and clone behavior. [[1]](diffhunk://#diff-6ea493f77408ea7e25ad4b19f2da630721ddd608e96db1ee26e833dc76e47014R1-R123) [[2]](diffhunk://#diff-d2b74c1abdb8daa15e3d4448f5824f61df79dbd3c5bdcda782f14c45baab2d49R1-R47)

**Dependency and import updates:**

- Added imports for `encoding/json` and `gopkg.in/yaml.v3` in relevant files to support the new serialization features. [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83R6-R8) [[2]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR5-R10) [[3]](diffhunk://#diff-6800f18f380946d94dd739a46e1ce72d445cf99d55bc2e42fb5c58b43284326bR4-R5)

These changes make it much easier and safer to persist and restore workflow state, with clear documentation and robust tests to guide usage.

### Related Issues

* Closes #74 
* Closes #76 
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: state_namespaced.go [hunk 1]
```diff
@@ -1,41 +1,40 @@
...
 // Local returns a view of the local namespace.
 func (n *SyncNamespacedStateBag) Local() StateBag {
-	n.localOnce.Do(func() {
-		if n.local == nil {
-			n.local = &SyncStateBag{}
-		}
-	})
 	n.mu.RLock()
 	defer n.mu.RUnlock()
 	return n.local
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply state_namespaced.go]: we removed lazy initialization
